server/src/handlers/users.rs

```diff
@@ -7,61 +7,262 @@ use axum::{
 use futures::TryStreamExt;
 use mongodb::{
     bson::oid::ObjectId,
-    results::{DeleteResult, InsertOneResult, UpdateResult},
+    results::{DeleteResult, UpdateResult},
 };
+use serde::{Deserialize, Serialize};
 
+use crate::auth::middleware::{AdminAuth, AuthorizationError, SelfOrAdminAuth};
+use crate::models::Token;
 use crate::{
+    auth,
     db_connection::DBClient,
     error::CheeseError,
     models::{User, UserList},
 };
 
 pub async fn fetch_users(
     Extension(db_client): Extension<DBClient>,
+    _auth: AdminAuth,
 ) -> Result<Json<UserList>, CheeseError> {
     let cursor = db_client.user_repo.get_all_users().await?;
-    Ok(Json(UserList {
-        users: cursor.try_collect().await?,
-    }))
+
+    let mut users: Vec<User> = cursor.try_collect().await?;
+    for u in &mut users {
+        u.hashed_password = String::new();
+    }
+
+    Ok(Json(UserList { users }))
 }
 
 pub async fn fetch_user(
     Path(id): Path<String>,
     Extension(db_client): Extension<DBClient>,
+    auth: SelfOrAdminAuth,
 ) -> Result<Json<Option<User>>, CheeseError> {
-    let res = db_client
-        .user_repo
-        .read_user(ObjectId::from_str(&id)?)
-        .await?;
-    Ok(Json(res))
-}
+    let id = ObjectId::from_str(&id)?;
+
+    auth.user_by_id(&id)?;
+
+    let mut res = db_client.user_repo.get_user_by_id(id).await?;
+
+    if let Some(ref mut user) = res {
+        user.hashed_password = String::new();
+    }
 
-pub async fn push_user(
-    Json(recipe): Json<User>,
-    Extension(db_client): Extension<DBClient>,
-) -> Result<Json<InsertOneResult>, CheeseError> {
-    let res = db_client.user_repo.create_user(recipe).await?;
     Ok(Json(res))
 }
 
 pub async fn update_user(
     Path(id): Path<String>,
-    Json(recipe): Json<User>,
+    Json(user): Json<User>,
     Extension(db_client): Extension<DBClient>,
+    auth: SelfOrAdminAuth,
 ) -> Result<Json<UpdateResult>, CheeseError> {
-    let obj_id = ObjectId::from_str(&id)?;
-    let res = db_client.user_repo.update_user(obj_id, recipe).await?;
+    let id = ObjectId::from_str(&id)?;
+
+    let req_user = auth.user_by_id(&id)?;
+
+    // if the requesting user is not an admin, they can not set themselves to be admin
+    if user.admin && !req_user.admin {
```

Comment:
we also need to check that the user isn't updating someone else, unless they are admin.

Reply:
Ah I missed auth was passed in as a parameter

Reply:
I don't see a `user.id == req_user.id` check? Or am I missing something else?
The JSON could still contain a malicious user_id that's different fram the Path parameter right?

Reply:
Mostly this is about what `.replace_one()` will do if the id in the object and the id in the filter don't match I think

Reply:
The easiest may just be to write a test (if there isn't one already).

Reply:
fixed in latest commit (and added a test)